transformer_engine/paddle/layer/attention.py

```diff
@@ -0,0 +1,458 @@
+                                               ctx.qkv_layout, ctx.attn_bias_type,
+                                               ctx.attn_mask_type)
+
+        # if no_bias, return dqkv
+        if ctx.attn_bias_type == "no_bias":
+            return (dqkv, None, None)
+        # else, return (dqkv, dbias)
+        return (dqkv, None, rest[0], None)
+
+
+class FusedAttnFuncPackedKV(paddle.autograd.PyLayer):
+    """Function for FusedAttention with packed KV input"""
+
+    @staticmethod
+    def forward(ctx, q, kv, cu_seqlens_q, cu_seqlens_kv, attn_bias, rng_state, max_seqlen_q,
+                max_seqlen_kv, attn_scale, qkv_dtype, dropout_p, set_zero, qkv_layout,
+                attn_bias_type, attn_mask_type, is_training):
+        out, aux_ctx_tensors = fused_attn_fwd_kvpacked(q, kv, cu_seqlens_q, cu_seqlens_kv,
+                                                       rng_state, is_training, max_seqlen_q,
+                                                       max_seqlen_kv, qkv_dtype, attn_bias,
+                                                       attn_scale, dropout_p, set_zero, qkv_layout,
+                                                       attn_bias_type, attn_mask_type)
+
+        ctx.save_for_backward(q, kv, out, cu_seqlens_q, cu_seqlens_kv, rng_state, aux_ctx_tensors)
+        ctx.max_seqlen_q = max_seqlen_q
+        ctx.max_seqlen_kv = max_seqlen_kv
+        ctx.qkv_dtype = qkv_dtype
+        ctx.attn_scale = attn_scale
+        ctx.dropout_p = dropout_p
+        ctx.set_zero = set_zero
+        ctx.qkv_layout = qkv_layout
+        ctx.attn_bias_type = attn_bias_type
+        ctx.attn_mask_type = attn_mask_type
+
+        return out
+
+    @staticmethod
+    def backward(ctx, d_out):
+        q, kv, out, cu_seqlens_q, cu_seqlens_kv, rng_state, aux_ctx_tensors = ctx.saved_tensor()
+        dq, dkv, *rest = fused_attn_bwd_kvpacked(q, kv, cu_seqlens_q, cu_seqlens_kv, rng_state, out,
+                                                 d_out, aux_ctx_tensors, ctx.max_seqlen_q,
+                                                 ctx.max_seqlen_kv, ctx.qkv_dtype, ctx.attn_scale,
+                                                 ctx.dropout_p, ctx.set_zero, ctx.qkv_layout,
+                                                 ctx.attn_bias_type, ctx.attn_mask_type)
+
+        # if no_bias, return dq, dkv
+        if ctx.attn_bias_type == "no_bias":
+            return (dq, dkv, None, None, None)
+        # else, return (dq, dkv, dbias)
+        return (dq, dkv, None, None, rest[0], None)
+
+
+class DotProductAttention(paddle.nn.Layer):
+    """Dot Product Attention Layer
+    BMM1 -> softmax + dropout -> BMM2
+    """
+
+    def __init__(self,
+                 norm_factor: int,
+                 attention_dropout: float = 0.0,
+                 attn_mask_type: str = "causal",
+                 attention_type: str = "self",
+                 backend: str = 'transformer_engine') -> None:
```

Comment:
I am not sure what you are saying.

This is TE/PyTorch:
```python
class DotProductAttention(...):

    if (condition):
        # calling cuDNN fused attention
    else:
        # calling flash_attention from hazyresearch
```

This is what we made for TE/PaddlePaddle for now:
```python
class DotProductAttention(...):

    if (condition):
        # transformer_engine backed, calling cuDNN fused attention
    else:
        # paddle backend, calling PaddlePaddle Primitives 
```

And what do you mean?

As far as I know, Baidu refers to flash_attention of hazyresearch and is developing its own version, but I am not sure the current status. Once complete, it will be in [PaddlePaddle repo](https://github.com/PaddlePaddle/Paddle) instead of [hazyresearch repo](https://github.com/Dao-AILab/flash-attention).

Reply:
The PyTorch behavior is more like:
```python
class DotProductAttention(...):
    if (condition1):
        # calling flash_attention from hazyresearch
    elif (condition2):
        # calling cuDNN fused attention
    else:
        # calling PyTorch primitives
```
This decision is made during the forward pass, not at construction time, and it can change depending on the sequence lengths. Also, it's possible for the attention to call PyTorch primitives while other Transformer layer components use TE kernels.

Reply:
So, there are 2 things.

1. add 3rd option - `calling flash_attention op (Baidu's version)`. 
   I need to sync with Baidu to know their status. This will take some time. Also, Paddle primitives should be automatically converted to flash by the Paddle deep learning compiler if the compiler is ready, so we might not need the 3rd option. I think cuDNN is still our top priority.
2.  move the logic of kernel selecting from constructor into forward. It is easy to do because PaddlePaddle supports dynamic graph mode. Although it has maximum flexibility, it will block CUDA Graph integration if kernel can be dynamically changed at runtime, won't it?

I think both 1. and 2. need more time to have better understanding. Considering schedule pressure, I hope this PR can be merged first, then 1. and 2. can be follow up in the future.

If I misunderstand anything, please correct me.
Thanks

Reply:
Yea, I think we can defer until later. Most of the backend logic in the PyTorch attention module is related to Flash Attention:
https://github.com/NVIDIA/TransformerEngine/blob/e4f9e7673b3b06d303164943f4ca33f46e865437/transformer_engine/pytorch/attention.py#L1086-L1113
It seems that this logic changes rapidly, as the cuDNN team adds kernels or fixes bugs.